### PR TITLE
anisotropy

### PR DESCRIPTION
**Pull request by unalmis**
- [x] Generalizes the change of variable in the singular kernel integrals to improves convergence of the integration. Before the integration region was an NFP elongated ellipse which will slow down convergence. That is changed here so that the region can be chosen to be a circle in real space to account for the affect of the the aspect ratio of the device has on the anisotropy of the grid.
- [ ] https://github.com/PlasmaControl/DESC/pull/1440/files#r1934449827
- [x] Change default chunk size for `virtual_casing_biot_savart` and free boundary
- [ ] Add option to account for anisotropy of grid in real space with local information

This PR must be squash merged.

### CODE REVIEW

**Comment by unalmis** [comment]
Thought I would separate this from the other branch

**Comment by dpanici** [comment]
Does this have a measurable impact on convergence rate compared to making it a circle in just (theta, zeta), not in real space?

**Comment by f0uriest** [comment]
Still looking into aspect ratio and grid anisotropy, but I've run a bunch of scans varying `s`, `q`, `Nu=Nv` etc for a number of equilibria. 

image
image
image
image
image
image
image
image


Main takeaways:
- choosing `s=q` seems to give the best performance
- cheaper to increase `q` than `Nu,Nv` but very little benefit to making `q >~ Nu/2`

**Comment by dpanici** [comment]
@dpanici run NFP axisym test here too

File: desc/objectives/_free_boundary.py [hunk 1]
```diff
@@ -354,10 +354,13 @@ class BoundaryError(_Objective):
         Equilibrium that will be optimized to satisfy the Objective.
     field : MagneticField
         External field produced by coils.
-    s : int
+    st : int
```
Comment (f0uriest) [inline desc/objectives/_free_boundary.py:357]: I think we could leave this as just `s` but allow it to be a tuple of two ints. So if its just a single int, we use the same for both, otherwise we treat it as `st` and `sz`

**Comment by dpanici** [comment]
> @dpanici run NFP axisym test here too

image


```python

from functools import partial

import numpy as np
import pytest
from jax import grad
from matplotlib import pyplot as plt
from numpy.polynomial.chebyshev import chebinterpolate, chebroots
from numpy.polynomial.legendre import leggauss
from scipy import integrate
from scipy.interpolate import CubicHermiteSpline
from scipy.special import ellipe, ellipkm1
# from tests.test_interp_utils import _f_1d, _f_1d_nyquist_freq, _f_2d, _f_2d_nyquist_freq
# from tests.test_plotting import tol_1d

from desc.backend import jnp, vmap
from desc.basis import FourierZernikeBasis
from desc.equilibrium import Equilibrium
from desc.equilibrium.coords import get_rtz_grid
from desc.examples import get
from desc.grid import ConcentricGrid, Grid, LinearGrid, QuadratureGrid
from desc.integrals import (
    Bounce1D,
    Bounce2D,
    DFTInterpolator,
    FFTInterpolator,
    line_integrals,
    singular_integral,
    surface_averages,
    surface_integrals,
    surface_integrals_transform,
    surface_max,
    surface_min,
    surface_variance,
    virtual_casing_biot_savart,
)
from desc.integrals._bounce_utils import (
    _check_bounce_points,
    _get_extrema,
    bounce_points,
)
from desc.integrals._interp_utils import fourier_pts
from desc.integrals.basis import FourierChebyshevSeries
from desc.integrals.quad_utils import (
    automorphism_sin,
    bijection_from_disc,
    chebgauss1,
    chebgauss2,
    get_quadrature,
    grad_automorphism_sin,
    grad_bijection_from_disc,
    leggauss_lob,
    tanh_sinh,
)
from desc.integrals.singularities import (
    _get_default_params,
    _get_quadrature_nodes,
    _kernel_nr_over_r3,
)
from desc.integrals.surface_integral import _get_grid_surface
from desc.transform import Transform
from desc.utils import dot, errorif, safediv
from desc.integrals.singularities import best_ratio, heuristic_support_params

# @pytest.mark.unit
#     def test_singular_integral_greens_id(self):
"""Test high order singular integration using greens identity.

Any harmonic function can be represented as the sum of a single layer and double
layer potential:

Φ(r) = -1/2π ∫ Φ(r) n⋅(r-r')/|r-r'|³ da + 1/2π ∫ dΦ/dn 1/|r-r'| da

If we choose Φ(r) == 1, then we get

1 + 1/2π ∫ n⋅(r-r')/|r-r'|³ da = 0

So we integrate the kernel n⋅(r-r')/|r-r'|³ and can benchmark the residual.

"""
eq = Equilibrium(NFP=1)
Nu = [13, 13, 13, 19, 19, 25, 37]
Nv = [30, 45, 60, 90, 120, 150, 240]
ss = [13, 13, 13, 19, 19, 25, 37]
qs = [10, 12, 12, 16, 18, 20, 24]
es = [0.4, 2e-2, 3e-3, 5e-5, 4e-6, 1e-6, 1e-9]
errs = []
for i in range(len(Nu)):
    grid = LinearGrid(M=Nu[i] // 2, N=Nv[i] // 2, NFP=eq.NFP)
    interpolator = FFTInterpolator(grid, grid, ss[i], ss[i], qs[i])
    data = eq.compute(
        _kernel_nr_over_r3.keys + ["|e_theta x e_zeta|"], grid=grid
    )
    err = singular_integral(data, data, "nr_over_r3", interpolator, chunk_size=10)
    np.testing.assert_array_less(np.abs(2 * np.pi + err), es[i])
    errs.append(np.mean(np.abs(2 * np.pi + err)))

eq = Equilibrium(NFP=19)
errsNFP19 = []
for i in range(len(Nu)):
    grid = LinearGrid(M=Nu[i] // 2, N=Nv[i] // 2, NFP=eq.NFP)
    interpolator = FFTInterpolator(grid, grid, ss[i], ss[i], qs[i])
    data = eq.compute(
        _kernel_nr_over_r3.keys + ["|e_theta x e_zeta|"], grid=grid
    )
    err = singular_integral(data, data, "nr_over_r3", interpolator, chunk_size=10)
    # np.testing.assert_array_less(np.abs(2 * np.pi + err), es[i])
    errsNFP19.append(np.mean(np.abs(2 * np.pi + err)))

errsNFP19_with_default = []
for i in range(len(Nu)):
    grid = LinearGrid(M=Nu[i] // 2, N=Nv[i] // 2, NFP=eq.NFP)
    st, sz, q = _get_default_params(grid)
    interpolator = FFTInterpolator(grid, grid, st, sz, q)
    data = eq.compute(
        _kernel_nr_over_r3.keys + ["|e_theta x e_zeta|"], grid=grid
    )
    err = singular_integral(data, data, "nr_over_r3", interpolator, chunk_size=10)
    # np.testing.assert_array_less(np.abs(2 * np.pi + err), es[i])
    errsNFP19_with_default.append(np.mean(np.abs(2 * np.pi + err)))
errsNFP19_new_default = []
for i in range(len(Nu)):
    grid = LinearGrid(M=Nu[i] // 2, N=Nv[i] // 2, NFP=eq.NFP)
    ratio_data = eq.compute(
        ["|e_theta x e_zeta|", "e_theta", "e_zeta"], grid=grid
    )
    st, sz, q = heuristic_support_params(grid, best_ratio(ratio_data)[0])
    interpolator = FFTInterpolator(grid, grid, st, sz, q)
    data = eq.compute(
        _kernel_nr_over_r3.keys + ["|e_theta x e_zeta|"], grid=grid
    )
    err = singular_integral(data, data, "nr_over_r3", interpolator, chunk_size=10)
    # np.testing.assert_array_less(np.abs(2 * np.pi + err), es[i])
    errsNFP19_new_default.append(np.mean(np.abs(2 * np.pi + err)))


import matplotlib.pyplot as plt
plt.figure()
total_N = np.array(Nu) * np.array(Nv)
plt.plot(total_N, errs,".-",label="NFP=1")
plt.plot(total_N, errsNFP19,".-",label="NFP=19 same params as NFP=1")
plt.plot(total_N, errsNFP19_with_default,".-",label="NFP=19 using _default_params")
plt.plot(total_N, errsNFP19_new_default,".-",label="NFP=19 using heuristic")

plt.yscale("log")
plt.legend()
```

**Comment by dpanici** [comment]
image

**Comment by f0uriest** [comment]
Notebook with bunch of plots for stellarator cases:
https://gist.github.com/f0uriest/f38547f0427fff2fa80d66422e18d143

**Comment by dpanici** [comment]
image

might be that like Rory said, we just need to use similar settings we'd use for NFP=1, but reduce N by NFP. something like ``N = round(np.max([Nv/eq.NFP, 2]) // 2)`` which is what I did above (the orange line techiincally has the bottom axis divided by NFP as its true total_N_points.

**Comment by dpanici** [comment]
Make fix for heuristic params for axisymmetry, see if then freeb test is fixed for axisym

File: desc/integrals/singularities.py [hunk 2]
```diff
@@ -134,21 +140,56 @@ def _heuristic_support_params(grid):
...
     """
     assert grid.can_fft2
-    # TODO: Account for real space grid anisotropy by choosing
-    #   ratio = |e_zeta|/|e_theta|
-    #   ratio_avg = surface_average(ratio, jacobian=|e_theta x e_zeta|)
-    #   st/(sz*NFP) (at evaluation point i) = mean (ratio(i), ratio_avg).
-    #   Then we have ~circle in real space around each singular point.
-    #   The grid resolution can then still be chosen independently according
-    #   to the frequency content of R, Z, λ to net the best function approximation.
+    # This is simple and should improve convergence, but do later.
```
Comment (dpanici) [inline desc/integrals/singularities.py:143]: if there is an issue for this, add to the comment, and if not, make an issue for using the local ratio

File: desc/objectives/_free_boundary.py [hunk 3]
```diff
@@ -431,9 +435,12 @@ def __init__(
             target = 0
         self._source_grid = source_grid
         self._eval_grid = eval_grid
-        self._eval_grid_is_source_grid = source_grid == eval_grid
-        self._st = s
-        self._sz = kwargs.get("sz", s)
+        if isinstance(s, (tuple, list)):
+            self._st = s[0]
```
Comment (dpanici) [inline desc/objectives/_free_boundary.py:439]: passing in s as tuple manually here for test coverage would be good

**Comment by unalmis** [comment]
> image
> 
> might be that like Rory said, we just need to use similar settings we'd use for NFP=1, but reduce N by NFP. something like `N = round(np.max([Nv/eq.NFP, 2]) // 2)` which is what I did above (the orange line techiincally has the bottom axis divided by NFP as its true total_N_points.

Can we default to this when we are adding fake field periods?

File: desc/integrals/singularities.py [hunk 4]
```diff
@@ -104,28 +104,105 @@ def _eta(theta, zeta, theta0, zeta0, ht, hz, st, sz):
     return _chi(r)
 
 
-def _get_default_params(grid):
-    k = max(min(grid.num_theta, grid.num_zeta * grid.NFP), 2)
-    s = k - 1
-    q = k // 2 + int(jnp.sqrt(k))
+def _vanilla_params(grid):
```
Comment (f0uriest) [inline desc/integrals/singularities.py:107]: can you use these in the tests? Same for `_local_params`

**Review by f0uriest — APPROVED** [review]

**Review by dpanici — APPROVED** [review]
